misc/amia_paper_2024/stats.py

```diff
@@ -0,0 +1,183 @@
+            df = pd.read_csv(path)
+            if verbose:
+                print(f'Read {len(df)} records from {path} in n seconds: {(datetime.now() - t0).seconds}')
+    # Mods
+    df['client_ip'] = df['client'].apply(lambda x: x.split(':')[0])
+    # Filtering
+    df = preprocess_null_call_groups(df, verbose)
+    # Formatting
+    df = df.fillna('')  # fixes problem w/ .str.contains() ops
+    return df
+
+
+def preprocess_dev_ips(df: pd.DataFrame, verbose=True) -> pd.DataFrame:
+    """Preprocess data: remove developers"""
+    # Devs accessing dev/prod
+    developer_ips = {'Siggie': '216.164.48.98', 'Joe': '174.99.54.40'}
+    if verbose:
+        print('Filtered out n records based on IP addresses of each developer accesing dev/prod:')
+        for dev, ip in developer_ips.items():
+            df_i = df[df['client_ip'] == ip]
+            print(f'  - {dev}: {len(df_i)}')
+    df2 = df[~df['client_ip'].isin(list(developer_ips.values()))]
+    # Devs running locally
+    count_before = len(df2)
+    df2 = df2[df2['host'].isin(['dev', 'prod'])]
+    if verbose:
+        print('Filtered out n records based devs running locally: ',  count_before - len(df2))
+    return df2
+
+def preprocess_null_call_groups(df: pd.DataFrame, verbose=True) -> pd.DataFrame:
+    """Preprocess data: remove null api_call_group_id
+
+    These are cases from before we added this feature, or where we called the backend directly.
+
+    As of 2023/03/14 this doesn't have an effect. No NULL groups."""
+    df2 = df[~df['api_call_group_id'].isna()]
+    if verbose:
+        print(f'Filtered out n records based on null api_call_group_id: ', len(df) - len(df2))
+    return df2
+
+
+# TODO: utilize df_apiruns
+def summary_stats(
+    df_apiruns: pd.DataFrame, df_w_groups_filtered: pd.DataFrame, outpath: Union[str, Path] = None
+):
+    """Generate summary statistics for given datasets
+
+    :param df_apiruns: Includes full logs, even for groups that have erroneous -1 as the api_call_group_id. This is only
+     used for tallying unique IP addresses.
+    :param df_w_groups_filtered: This is a JOIN of apiruns and apiruns_grouped tables, with erroneous api_call_group_id
+     of -1 filtered out."""
+    # Get summary statistics
+    summary = {}
+    summary['Total log records'] = len(df_w_groups_filtered)
+    summary['Log sessions (api_call_group_id)'] = len(df_w_groups_filtered['api_call_group_id'].unique())
+    summary['IP addresses'] = len(df_apiruns['client_ip'].unique())
+    summary['Errors'] = df_w_groups_filtered[df_w_groups_filtered['result'].str.lower().str.contains('error')][
+        'api_call_group_id'].nunique()
+
+    # Value set combos
+    df_get_csets = df_w_groups_filtered[df_w_groups_filtered['api_call'] == 'get-csets']
+    summary['Value set combos'] = len(set([tuple(x) for x in df_get_csets['codeset_ids']]))
+
+    # todo: temp: analysis
+    #  - analyze uniqueness of get-csets calls within an API call group session
```

Comment:
## Value set combos
**How I'm ascertaining (for now)**: By unique calls on `get-csets`.
I did it this way in the interest of time. It is not a bad proxy for the truth., and may even be revealed to be the truth, after addressing data quality issues.

**Data quality & strength of this ascertainment method**: I did not expect this, but out of 2,944 `get-csets` calls, 30 `api_call_group_id` sessions where the same call (same combo of csets) was being called multiple times. I don't know if this is a programming error or an artefact of lag, or a gap in my understanding. There was only 1 case in which there were 2 `get-csets` calls in a single group which had different csets. All other sessions had only 1 call per session.

<details><summary>Details</summary>
<p>

See comments after variable names for values.

```py
    # Value set combos
    df_get_csets = df_w_groups_filtered[df_w_groups_filtered['api_call'] == 'get-csets']
    summary['Value set combos'] = len(set([tuple(x) for x in df_get_csets['codeset_ids']]))

    # todo: temp: analysis
    #  - analyze uniqueness of get-csets calls within an API call group session
    # diff_get_csets_vs_groups=35; i feel like this should be equal to some calc involving next 4 vars, but not sure
    # noinspection PyUnusedLocal
    diff_get_csets_vs_groups = len(df_get_csets) - len(df_get_csets['api_call_group_id'].unique())  # 35
    diff_codeset_ids_in_group__n_instances = 0  # 1
    diff_codeset_ids_in_group__n_calls = 0  # 2
    mult_get_cset_in_session__n_instances = 0  # 30
    mult_get_cset_in_session__n_calls = 0  # 64
    # todo: maybe convert to str in pre-processing instead, if need 'str' more than temporarily
    df_get_csets['codeset_ids'] = df_get_csets['codeset_ids'].astype(str)
    for group_id, group_data in df_get_csets.groupby('api_call_group_id'):
        if len(group_data) > 1 and len(group_data['codeset_ids'].unique()) > 1:
            diff_codeset_ids_in_group__n_instances += 1
            diff_codeset_ids_in_group__n_calls += len(group_data)
        elif len(group_data) > 1:
            mult_get_cset_in_session__n_instances += 1
            mult_get_cset_in_session__n_calls += len(group_data)
```

</p>
</details> 

**Why I chose this method:** I didn't have knowledge of what the unique ways in which an `api_call_group_id` session could be started. As far as I could see, they almost always started off with `get-csets`, and in cases where they did not, they were soon followed by a call to `get-csets`, almost always calling the sam set of IDs.

**Alternate solutions**: I suppose I could look at the first timestamp for each session, but I'm not sure it is always representative of the cset combos being primarily queried.